### ISSUE TITLE
Bootstrap installation of numpy in setup.py if it is not already installed

### ISSUE THREAD

**Issue by ziggi0703** [rootpy/root_numpy#174]
Hello,

I use root_numpy in a package of my own and want to make it possible to install it with pip. But the installation fails due to your import of numpy in the setup.py. 

Is it possible to not import the package dependency in the setup.py so I don't have to install numpy by hand before?

Cheers
ziggi

**Comment by kratsg** [comment]
I'm not 100% sure I understand, are you saying doing

```
pip install numpy
pip install root_numpy
```

doesn't work for you? I use https://virtualenvwrapper.readthedocs.org/en/latest/ to manage virtual environments so I can bundle up installations together.

**Comment by ziggi0703** [comment]
pip install numpy 
pip install root_numpy 

works fine. But this is not what I really want if I'm using pip. I just want to do 

pip install root_numpy

and numpy is then installed automatically.

**Comment by srstevenson** [comment]
@ziggi0703 I don't think that'll be possible, as `setup.py` needs to know where the numpy headers are in order [to compile `librootnumpy `](https://github.com/rootpy/root_numpy/blob/d388474ca360cb3906e50291fa522137479bd614/setup.py#L63-L73).

**Comment by kratsg** [comment]
Yeah, @srstevenson is correct, see here: http://nullege.com/codes/search/numpy.get_include and the relevant line here: https://github.com/rootpy/root_numpy/blob/d388474ca360cb3906e50291fa522137479bd614/setup.py#L70

I guess what can happen is providing a dependency on `numpy` explicitly by `install_requires` but leave the rest of the code intact. This would at least let `pip` or another appropriate python installer grab `numpy` first before `root_numpy`.

**Comment by ziggi0703** [comment]
@kratsg I think this will not solve the problem, as the dependencies given in the ```setup.py``` cannot be read by ```pip``` if ```numpy``` is not installed. 

But I think I will stick at
```
pip install numpy
pip install root_numpy
```

@srstevenson @kratsg  thanks for your answers

**Comment by kratsg** [comment]
I have an idea that I'll test and update you later if it fleshes out :)

**Comment by ziggi0703** [comment]
@kratsg thanks

**Comment by kratsg** [comment]
Hi @ziggi0703 , none of my ideas fleshed out. There's no easy way of getting this resolved. One of the main issues is that one can install numpy in a non-default location or non-local location, which means having the system find the headers is a pain and the only way to honestly find the location consistently is via `python -c "import numpy; numpy.get_include()"`. For some reason, even explicitly making pip install numpy as a dependency doesn't allow us to directly import unless pip is executed twice, one for numpy and one for root_numpy.

**Comment by ndawe** [comment]
This should in fact be possible. I'm looking at the setup in scikit-learn since they also have extension modules that depend on numpy, but you can pip install scikit-learn without first installing numpy (numpy will then be installed as a requirement).

**Comment by kratsg** [comment]
> This should in fact be possible. I'm looking at the setup in scikit-learn since they also have extension modules that depend on numpy, but you can pip install scikit-learn without first installing numpy (numpy will then be installed as a requirement).

False. (I use makevirtualenv for testing/environments).

```(soundcloud)Lord Stark:~$ pip list
decorator (3.4.0)
fudge (1.0.3)
networkx (1.9.1)
pip (1.5.6)
requests (2.4.3)
setuptools (3.6)
simplejson (3.6.5)
soundcloud (0.4.1)
wsgiref (0.1.2)
```

then run `pip install scikit-learn` then
```(soundcloud)Lord Stark:~$ pip list
decorator (3.4.0)
fudge (1.0.3)
networkx (1.9.1)
pip (1.5.6)
requests (2.4.3)
scikit-learn (0.15.2)
setuptools (3.6)
simplejson (3.6.5)
soundcloud (0.4.1)
wsgiref (0.1.2)
```

all good so far. Then this fails

```
(soundcloud)Lord Stark:~$ python -c "import sklearn"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/kratsg/.virtualenvs/soundcloud/lib/python2.7/site-packages/sklearn/__init__.py", line 38, in <module>
    from .base import clone
  File "/Users/kratsg/.virtualenvs/soundcloud/lib/python2.7/site-packages/sklearn/base.py", line 9, in <module>
    import numpy as np
ImportError: No module named numpy
```

**Comment by ndawe** [comment]
That is strange since installing scikit-learn should build the internal extensions and they depend on numpy. Testing now for myself...

**Comment by ndawe** [comment]
Yeah, installing scikit-learn without first installing numpy should result in this error during the pip install scikit-learn:

ImportError: No module named numpy.distutils.core

(I tested in a virtualenv)

You should also have hit this if in fact the environment in which you pip installed scikit-learn did not have access to numpy.

**Comment by ndawe** [comment]
So, it seems they don't have a solution to this either. At least root_numpy prints a helpful message:

```
        "numpy cannot be imported. numpy must be installed "
    
    RuntimeError: numpy cannot be imported. numpy must be installed prior to installing root_numpy

```

**Comment by ndawe** [comment]
See the discussion here:

http://stackoverflow.com/questions/19919905/how-to-bootstrap-numpy-installation-in-setup-py

**Comment by ndawe** [comment]
We could what's done in the scipy setup.py although it's a bit of a hack:

https://github.com/scipy/scipy/blob/master/setup.py

**Comment by kratsg** [comment]
The reason I'm not hitting it is because I installed from a wheel (pre-installed binary packaged up in an egg). I can force the no-wheel and get the warning.

**Comment by ndawe** [comment]
A solution is finally implemented in #292
